dockstore-client/src/main/java/io/dockstore/client/cli/nested/WesLauncher.java

```diff
@@ -0,0 +1,203 @@
+     *
+     * @param workflowClient The WorkflowClient
+     * @param workflowEntry The workflow entry (i.e. github.com/myRepo/myWorkflow:version)
+     * @param workflow The workflow object we are creating a TRS url for
+     * @return A string representing a TRS URL for an entry
+     */
+    public static String combineTrsUrlComponents(WorkflowClient workflowClient, String workflowEntry, Workflow workflow) {
+        ApiClient client = workflowClient.getClient().getGa4Ghv20Api().getApiClient();
+
+        // Entries are passed in the form {PATH}:{VERSION} or {PATH}
+        final String[] pathAndVersion = workflowEntry.split(":");
+        final String path = pathAndVersion[0];
+
+        // Calculate the values needed to supply a TRS URL
+        final String basePath = workflowClient.getClient().getGa4Ghv20Api().getApiClient().getBasePath();
+        final String versionId = workflowClient.getVersionID(workflowEntry);
+        final String entryId = workflowClient.getTrsId(path);
+        final String type = "PLAIN_" + workflow.getDescriptorType().getValue();
+
+        // Escape each of the URL path values
+        final String escapedId = client.escapeString(entryId);
+        final String escapedVersionId = client.escapeString(versionId);
+        final String escapedType = client.escapeString(type);
+        final String escapedRelativePath = client.escapeString(workflow.getWorkflowPath());
+
+        // Return the TRS URL for a given entry
+        return MessageFormat.format("{0}/ga4gh/trs/v2/tools/{1}/versions/{2}/{3}/descriptor/{4}",
+            basePath,
+            escapedId,
+            escapedVersionId,
+            escapedType,
+            escapedRelativePath);
+    }
+
+    /**
+     * Attempts to load a file from the provided path. Errors out if the file doesn't exist.
+     *
+     * @param filePath Path to a file or null
+     * @return An Optional File object, this may be empty if the filepath is null
+     */
+    public static Optional<File> fetchFile(String filePath) {
+        if (filePath == null) {
+            return Optional.empty();
+        }
+
+        // Verify the file path exists
+        Path path = Paths.get(filePath);
+        if (!Files.exists(path)) {
+            errorMessage(MessageFormat.format("Unable to locate file: {0}", filePath), CLIENT_ERROR);
+        }
+
+        // TODO: Handle path expansions? i.e. '~/my/file.txt'
+        return Optional.of(new File(filePath));
+    }
+
+    /**
+     * Given a list of string paths, this attempts to convert it to a list of files.
+     *
+     * @param filePaths A list of paths that correspond to files that need to be attached to the WES request, this may be null
+     * @return A list of File objects, which may be empty
+     */
+    public static List<File> fetchFiles(List<String> filePaths) {
+
+        // Return an empty list if no attachments were passed
+        if (filePaths == null) {
+            return new ArrayList<>();
```

Comment:
Tip, fine the way you have it, but you can return `Collections.emptyList()` -- it's more for the use case where a method is called thousands or millions of times, so that you're not creating objects that need to be garbage collected, but since this is called once, it doesn't matter, but just for future reference.

Reply:
Interesting, I hadn't considered that case, thanks for the pointer.

Updated: https://github.com/dockstore/dockstore-cli/pull/100/commits/09bf7c4684c788157d1b3e3f8838a7cca44b246c